charmhelpers/contrib/network/ovs/ovsdb.py

```diff
@@ -28,74 +28,183 @@ class SimpleOVSDB(object):
+            'datapath',
+            'flow_sample_collector_set',
+            'flow_table',
+            'ipfix',
+            'interface',
+            'manager',
+            'mirror',
+            'netflow',
+            'open_vswitch',
+            'port',
+            'qos',
+            'queue',
+            'ssl',
+            'sflow',
+        ),
+        'ovn-nbctl': (
+            'acl',
+            'address_set',
+            'connection',
+            'dhcp_options',
+            'dns',
+            'forwarding_group',
+            'gateway_chassis',
+            'ha_chassis',
+            'ha_chassis_group',
+            'load_balancer',
+            'load_balancer_health_check',
+            'logical_router',
+            'logical_router_policy',
+            'logical_router_port',
+            'logical_router_static_route',
+            'logical_switch',
+            'logical_switch_port',
+            'meter',
+            'meter_band',
+            'nat',
+            'nb_global',
+            'port_group',
+            'qos',
+            'ssl',
+        ),
+        'ovn-sbctl': (
+            'address_set',
+            'chassis',
+            'connection',
+            'controller_event',
+            'dhcp_options',
+            'dhcpv6_options',
+            'dns',
+            'datapath_binding',
+            'encap',
+            'gateway_chassis',
+            'ha_chassis',
+            'ha_chassis_group',
+            'igmp_group',
+            'ip_multicast',
+            'logical_flow',
+            'mac_binding',
+            'meter',
+            'meter_band',
+            'multicast_group',
+            'port_binding',
+            'port_group',
+            'rbac_permission',
+            'rbac_role',
+            'sb_global',
+            'ssl',
+            'service_monitor',
+        ),
+    }
+
+    def __init__(self, tool):
+        """SimpleOVSDB constructor.
 
         :param tool: Which tool with database commands to operate on.
                      Usually one of `ovs-vsctl`, `ovn-nbctl`, `ovn-sbctl`
         :type tool: str
-        :param table: Which table to operate on
-        :type table: str
         """
-        if tool not in ('ovs-vsctl', 'ovn-nbctl', 'ovn-sbctl'):
+        if tool not in self._tool_table_map:
             raise RuntimeError(
-                "tool must be one of 'ovs-vsctl', 'ovn-nbctl', 'ovn-sbctl'")
+                'tool must be one of "{}"'.format(self._tool_table_map.keys()))
         self.tool = tool
-        self.tbl = table
-
-    def _find_tbl(self, condition=None):
-        """Run and parse output of OVSDB `find` command.
-
-        :param condition: An optional RFC 7047 5.1 match condition
-        :type condition: Optional[str]
-        :returns: Dictionary with data
-        :rtype: Iterator[Dict[str, ANY]]
+        self._tool = tool
+        self._tableobject = {}
+
+    def __getattr__(self, table):
+        if table not in self._tool_table_map[self._tool]:
```

Comment:
It took me a little while to work it out, but this class is building a set of singleton `Table` objects on demand?  Whilst that would be useful if the objects held state (which they don't seem to?), all this is doing is keeping the objects around after they have been used until the parent `SimpleOVSDB` goes away ... i.e goes out of scope?  Is this what is intended?

If not, why memoise the object, and instead just hand out a new `Table` object after it's been verified that the table sting is valid?

Reply:
I guess you caught me red handed in performing premature optimization! Gone. :+1: